demos/coreMQTT_Agent/simple_sub_pub_demo.c

```diff
@@ -0,0 +1,560 @@
+    BaseType_t xCommandAcknowledged = pdFALSE;
+    uint32_t ulSubscribeMessageID;
+    MQTTAgentSubscribeArgs_t xSubscribeArgs;
+    MQTTSubscribeInfo_t xSubscribeInfo;
+    static int32_t ulNextSubscribeMessageID = 0;
+    MQTTAgentCommandContext_t xApplicationDefinedContext = { 0 };
+    MQTTAgentCommandInfo_t xCommandParams = { 0 };
+
+    /* Create a unique number of the subscribe that is about to be sent.  The number
+     * is used as the command context and is sent back to this task as a notification
+     * in the callback that executed upon receipt of the subscription acknowledgment.
+     * That way this task can match an acknowledgment to a subscription. */
+    xTaskNotifyStateClear( NULL );
+    taskENTER_CRITICAL();
+    {
+        ulNextSubscribeMessageID++;
+        ulSubscribeMessageID = ulNextSubscribeMessageID;
+    }
+    taskEXIT_CRITICAL();
+
+    /* Complete the subscribe information.  The topic string must persist for
+     * duration of subscription! */
+    xSubscribeInfo.pTopicFilter = pcTopicFilter;
+    xSubscribeInfo.topicFilterLength = ( uint16_t ) strlen( pcTopicFilter );
+    xSubscribeInfo.qos = xQoS;
+    xSubscribeArgs.pSubscribeInfo = &xSubscribeInfo;
+    xSubscribeArgs.numSubscriptions = 1;
+
+    /* Complete an application defined context associated with this subscribe message.
+     * This gets updated in the callback function so the variable must persist until
+     * the callback executes. */
+    xApplicationDefinedContext.ulNotificationValue = ulNextSubscribeMessageID;
+    xApplicationDefinedContext.xTaskToNotify = xTaskGetCurrentTaskHandle();
+    xApplicationDefinedContext.pArgs = ( void * ) &xSubscribeArgs;
+
+    xCommandParams.blockTimeMs = mqttexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = prvSubscribeCommandCallback;
+    xCommandParams.pCmdCompleteCallbackContext = ( void * ) &xApplicationDefinedContext;
+
+    /* Loop in case the queue used to communicate with the MQTT agent is full and
+     * attempts to post to it time out.  The queue will not become full if the
+     * priority of the MQTT agent task is higher than the priority of the task
+     * calling this function. */
+    LogInfo( ( "Sending subscribe request to agent for topic filter: %s with id %d",
+               pcTopicFilter,
+               ( int ) ulSubscribeMessageID ) );
+
+    do
+    {
+        /* TODO: prvIncomingPublish as publish callback. */
+        xCommandAdded = MQTTAgent_Subscribe( &xGlobalMqttAgentContext,
+                                             &xSubscribeArgs,
+                                             &xCommandParams );
+    } while( xCommandAdded != MQTTSuccess );
+
+    /* Wait for acks to the subscribe message - this is optional but done here
+     * so the code below can check the notification sent by the callback matches
+     * the ulNextSubscribeMessageID value set in the context above. */
+    xCommandAcknowledged = prvWaitForCommandAcknowledgment( NULL );
```

Comment:
Why has `NULL` been passed to the function if the goal is the match the ulNextSubscribeMessaeID value?

Reply:
This may have been since it's not necessary for this specific demo since there's only one subscribe per task, but I updated this